lib/semigrp.gi

```diff
@@ -542,7 +542,25 @@ InstallMethod( AsSemigroup,
     "generic method for collections",
     [ IsCollection ],
     function ( D )
-    local   S,  L;
+    local L, S, i, j, k;
+
+    # Must check associativity
+    if ApplicableMethod(IsAssociative, [D]) <> fail then
+      # In case there are better methods/known attributes for associativity
+      if not IsAssociative(D) then
+        return fail;
+      fi;
+    else
+      for i in D do
+        for j in D do
+          for k in D do
+            if (i * j) * k <> i * (j * k) then
```

Comment:
Would it perhaps make sense to install this as a generic fallback method for  IsAssociative? Then the code in this function could be considerably simplified.

Reply:
I think there is an argument to be made the `IsAssociative` (and also `IsCommutative`, and maybe more) also make sense for arbitrary *subsets* of magmas. If one accepts this point of view, it would make sense to install methods for both for `IsMultiplicativeElementCollection`. One could also have `InstallTrueMethod( IsAssociative, IsAssociativeElementCollection );`

However, I am not sure if this has a risk of breaking something? I can't think of anything, though...

Wonder if @ThomasBreuer has some thoughts on this...

Reply:
It seems sensible to me... I'll give it a go eventually 🙂